### ISSUE TITLE
Common Installation Problems

### ISSUE THREAD

**Issue by awni** [ml-explore/mlx#200]
The docs have more information on [troubleshooting installation](https://ml-explore.github.io/mlx/build/html/install.html).

Two of the more common issues are described below (as well as in the docs).

# Installing from PyPI

A common problem is `pip install mlx` does not find a matching distribution:

```
ERROR: Could not find a version that satisfies the requirement mlx (from versions: none)
ERROR: No matching distribution found for mlx
```

If this happens you are probably using a non-native Python. The output of:

```
python -c "import platform; print(platform.processor())"
```

Should be `arm` not `i386`. You can fix this by [using an environment with a native Python using Conda](https://stackoverflow.com/questions/65415996/how-to-specify-the-architecture-or-platform-for-a-new-conda-environment-apple). Another option is to simply reinstall Conda and setup a new environment.

# Installing from Source

You see the following error when you try to build:

```
error: unable to find utility "metal", not a developer tool or in PATH
```

To fix this, first make sure you have Xcode installed:

```
xcode-select --install
```

Then set the active developer directory:

```
sudo xcode-select --switch /Applications/Xcode.app/Contents/Developer
```

**Comment by ariaattar** [comment]
<img width="526" alt="Screen Shot 2023-12-21 at 3 50 06 PM" src="https://github.com/ml-explore/mlx/assets/72599441/aefdbde0-7856-44e9-902a-c7aeff0ba042">

Having the same issue, platform is outputting `arm` is there something else I should check?

**Comment by awni** [comment]
operating system?

**Comment by matthewfernst** [comment]
Hi there, I'm getting the following error during `make install` is there something silly I'm missing? I was following the steps mentioned [here](https://ml-explore.github.io/mlx/build/html/install.html). It works if I do `sudo make install`, but when I try running a `isnan` that I'm implementing I'm getting this error `[metal::Device] Unable to load kernel ssnaneqint32`. I'm guessing something is wrong with the install? Thank you!

```bash
CMake Error at mlx/backend/metal/kernels/cmake_install.cmake:41 (file):
  file INSTALL cannot copy file
  "/Users/matthewernst/Developer/AIML/mlx/build/mlx/backend/metal/kernels//mlx.metallib"
  to "/usr/local/lib/mlx.metallib": Permission denied.
Call Stack (most recent call first):
  mlx/backend/metal/cmake_install.cmake:42 (include)
  mlx/cmake_install.cmake:57 (include)
  cmake_install.cmake:42 (include)
```

**Comment by aakashvardhan** [comment]
Is this not applicable for pyenv?

**Comment by awni** [comment]
I'm not sure what you mean. You should be able to use mlx inside a pyenv.

**Comment by oldfishoh** [comment]
I resolved the issue by reinstalling conda, and successfully installed it using pip install chat-with-mlx. However, how should I run it?

**Comment by awni** [comment]
> However, how should I run it?

To run chat-with-mlx I would [check the docs there](https://github.com/qnguyen3/chat-with-mlx). That is a third party repo. If you have issues with it, I would encourage you to file an issue in that repo.

**Comment by faisalmemon** [comment]
I had an installation problem with my setup which wasn't based on miniconda (#1846).  The step where you created an editable package failed because it got confused between python 3.9 and 3.13.0.  To overcome the issue I switched to miniconda. Starting a system with zsh for the shell and using the brew package manager, I was able to do the following:
```
cd mlx
rm -rf *
git checkout -- .
brew install --cask miniconda
conda init "$(basename "${SHELL}")"
source ~/.zshrc
conda create --name myenv python=3.13.0
conda activate myenv
pip install --upgrade pip
CMAKE_BUILD_PARALLEL_LEVEL=8 pip install ".[dev]"
CMAKE_BUILD_PARALLEL_LEVEL=8 pip install -e ".[dev]"
```

And this makes the `-e` option work for me with python 3.13.0.  It previously only worked for 3.9.x without miniconda.

It was probably a setup glitch on my system, but I am reporting this here in case it is a common issue or it helps others.

**Comment by 13rac1** [comment]
I have Python 3.14 installed, but need to use Python 1.13:
```zsh
% pip3 install mlx
ERROR: Could not find a version that satisfies the requirement mlx (from versions: none)
ERROR: No matching distribution found for mlx
% uv venv --python 3.13 .venv
Using CPython 3.13.9 interpreter at: /opt/homebrew/opt/python@3.13/bin/python3.13
Creating virtual environment at: .venv
Activate with: source .venv/bin/activate
% source .venv/bin/activate
% uv pip install mlx
Resolved 2 packages in 87ms
Prepared 1 package in 48ms
Installed 2 packages in 12ms
 + mlx==0.29.3
 + mlx-metal==0.29.3
```

**Comment by awni** [comment]
Indeed we don't yet have a build for Python 3.14. It just entered stable a week or two ago. We'll update the next release to include it. For now you'll have to build from source or use Python 3.13.

**Comment by 13rac1** [comment]
@awni Yes, I see that. The error message from `pip` is useless, so I posted a clear solution.
